WebTransport should be able to cancel reads or writes for a stream

AnkshitJain opened the pull request:
#### c9f3e1d327891574fc0d8b5fa734f0e7153a6ce3
<pre>
WebTransport should be able to cancel reads or writes for a stream
<a href="https://bugs.webkit.org/show_bug.cgi?id=303152">https://bugs.webkit.org/show_bug.cgi?id=303152</a>
<a href="https://rdar.apple.com/160150325">rdar://160150325</a>

Reviewed by Alex Christensen.

This PR plumbs sending and receiving RESET_STREAM_AT and STOP_SENDING signals on the streams.

The sending flow does not take the stream error code into account right now.
This is tracked by existing FIXMEs in `WebTransportReceiveStreamSource::doCancel` and `WebTransportSendStreamSink::error`.

It also fixes the cleanup procedure for WebTransport, using the correct exception to terminate the readable and writable streams.

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/WebTransport.mm

* Source/WebCore/Modules/streams/ReadableStreamSource.cpp:
(WebCore::ReadableStreamSource::error):
* Source/WebCore/Modules/streams/ReadableStreamSource.h:
* Source/WebCore/Modules/streams/WritableStream.cpp:
(WebCore::WritableStream::errorIfPossible):
* Source/WebCore/Modules/streams/WritableStream.h:
* Source/WebCore/Modules/webtransport/DatagramByteSource.cpp:
(WebCore::DatagramByteSource::error):
* Source/WebCore/Modules/webtransport/DatagramByteSource.h:
* Source/WebCore/Modules/webtransport/DatagramSource.cpp:
(WebCore::DatagramDefaultSource::error):
* Source/WebCore/Modules/webtransport/DatagramSource.h:
* Source/WebCore/Modules/webtransport/WebTransport.cpp:
(WebCore::WebTransport::receiveBidirectionalStream):
(WebCore::WebTransport::streamReceiveError):
(WebCore::WebTransport::streamSendError):
(WebCore::WebTransport::cleanup):
(WebCore::WebTransport::createBidirectionalStream):
(WebCore::WebTransport::createUnidirectionalStream):
(WebCore::WebTransport::didFail):
* Source/WebCore/Modules/webtransport/WebTransport.h:
* Source/WebCore/Modules/webtransport/WebTransportReceiveStreamSource.cpp:
(WebCore::WebTransportReceiveStreamSource::receiveError):
* Source/WebCore/Modules/webtransport/WebTransportReceiveStreamSource.h:
* Source/WebCore/Modules/webtransport/WebTransportSessionClient.h:
* Source/WebCore/Modules/webtransport/WorkerWebTransportSession.cpp:
(WebCore::WorkerWebTransportSession::didFail):
(WebCore::WorkerWebTransportSession::streamReceiveError):
(WebCore::WorkerWebTransportSession::streamSendError):
* Source/WebCore/Modules/webtransport/WorkerWebTransportSession.h:
* Source/WebCore/PAL/pal/spi/cocoa/NetworkSPI.h:
* Source/WebCore/bindings/js/InternalWritableStream.cpp:
(WebCore::InternalWritableStream::errorIfPossible):
* Source/WebCore/bindings/js/InternalWritableStream.h:
* Source/WebCore/bindings/js/ReadableStreamDefaultController.cpp:
(WebCore::ReadableStreamDefaultController::error):
* Source/WebCore/bindings/js/ReadableStreamDefaultController.h:
* Source/WebKit/Configurations/AllowedSPI.toml:
* Source/WebKit/NetworkProcess/cocoa/NetworkSoftLink.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkSoftLink.mm:
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.cpp:
(WebKit::NetworkTransportSession::streamReceiveError):
(WebKit::NetworkTransportSession::streamSendError):
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.h:
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportStream.h:
* Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
(WebKit::NetworkTransportSession::setupConnectionHandler):
(WebKit::NetworkTransportSession::createStream):
* Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportStreamCocoa.mm:
(WebKit::NetworkTransportStream::sendBytes):
(WebKit::NetworkTransportStream::receiveLoop):
(WebKit::NetworkTransportStream::cancel):
(WebKit::NetworkTransportStream::cancelReceive):
(WebKit::NetworkTransportStream::cancelSend):
(WebKit::NetworkTransportStream::setErrorCodeForStream): Deleted.
* Source/WebKit/WebProcess/Network/WebTransportSession.cpp:
(WebKit::WebTransportSession::streamReceiveError):
(WebKit::WebTransportSession::streamSendError):
(WebKit::WebTransportSession::didFail):
* Source/WebKit/WebProcess/Network/WebTransportSession.h:
* Source/WebKit/WebProcess/Network/WebTransportSession.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebTransport.mm:
(TestWebKitAPI::TEST(WebTransport, NetworkProcessCrash)):

Canonical link: <a href="https://commits.webkit.org/303774@main">https://commits.webkit.org/303774@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/25fe9cc48f173258b39622de43cd6ed0dcbf058a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/133407 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/5908 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/44545 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/140963 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/85454 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/135277 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/6421 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/5773 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102050 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/69481 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/136354 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/4551 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/119581 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/82844 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/4428 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2025 "Passed tests") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/113531 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/37691 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/143609 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/5578 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/38278 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/110425 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/5660 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/4787 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110607 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4290 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/115840 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/59328 "The change is no longer eligible for processing.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20648 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/5633 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/34166 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/5479 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/69085 "Built successfully") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/5722 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/5589 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/Modules/streams/ReadableStreamSource.cpp:
```diff
@@ -77,6 +77,11 @@ void ReadableStreamSource::cancel(JSC::JSValue)
 
+void ReadableStreamSource::error(JSC::JSGlobalObject& globalObject, JSC::JSValue value)
+{
+    m_controller->error(globalObject, value);
```
if (m_controller)

achristensen07 commented on Source/WebCore/Modules/webtransport/WebTransport.h:
```diff
@@ -151,6 +154,7 @@ class WebTransport : public WebTransportSessionClient, public ActiveDOMObject {
     const Ref<WebTransportReceiveStreamSource> m_receiveStreamSource;
     const Ref<WebTransportBidirectionalStreamSource> m_bidirectionalStreamSource;
     HashMap<WebTransportStreamIdentifier, Ref<WebTransportReceiveStreamSource>> m_readStreamSources;
+    HashMap<WebTransportStreamIdentifier, Ref<WebTransportSendStream>> m_writeStreams;
```
It seems like we should remove m_sendStreams when we do this and replace its logic with this.

AnkshitJain replied:
`sendStreams` is an [internal slot](https://www.w3.org/TR/webtransport/#dom-webtransport-sendstreams-slot) for WebTransport. We will likely need a structure like ListHashMap for this change.

achristensen07 commented on Source/WebCore/Modules/webtransport/WebTransport.cpp:
```diff
@@ -413,25 +457,50 @@ void WebTransport::close(WebTransportCloseInfo&& closeInfo)
-    m_incomingUnidirectionalStreams->cancel(Exception { ExceptionCode::NetworkError });
+        stream->errorIfPossible(jsDOMGlobalObject, jsException);
+
+    std::exchange(m_writeStreams, { });
```
In cases where we don't use the return value of std::exchange, it's probably better to just write m_X = { };
This needs a bit of rewriting anyways because of m_sendStreams's duplicate nature.

achristensen07 commented on Source/WebCore/Modules/webtransport/WebTransport.cpp:
```diff
@@ -413,25 +457,50 @@ void WebTransport::close(WebTransportCloseInfo&& closeInfo)
+    auto* globalObject = context->globalObject();
+    if (!globalObject)
+        return;
+    auto& jsDOMGlobalObject = *JSC::jsCast<JSDOMGlobalObject*>(globalObject);
```
Let's use jsDynamicCast or jsSecureCast just to be a little safer.

achristensen07 commented on Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
```diff
@@ -504,19 +504,35 @@ static String joinProtocolStrings(const Vector<String>& protocols)
-            }
-            Ref stream = NetworkTransportStream::create(*protectedThis.get(), std::exchange(inboundConnection, nullptr).get(), NetworkTransportStreamType::Bidirectional);
+            bool unidirectional = nw_webtransport_metadata_get_is_unidirectional(metadata.get());
+            NetworkTransportStreamType streamType = unidirectional ? NetworkTransportStreamType::IncomingUnidirectional : NetworkTransportStreamType::Bidirectional;
```
nit: we can use auto for the first NetworkTransportStreamType because it is very clear what the type is.

achristensen07 commented on Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportStreamCocoa.mm:
```diff
@@ -85,21 +85,25 @@
-                break;
-            case NetworkTransportStreamState::WriteClosed:
-                RELEASE_ASSERT_NOT_REACHED();
+        if (protectedThis) {
```
nit: it would reduce unnecessary indentation to have an early return:
if (!protectedThis)
    return completionHandler(std::nullopt);

achristensen07 commented on Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportStreamCocoa.mm:
```diff
@@ -114,8 +118,10 @@
         if (error) {
-            if (!(nw_error_get_error_domain(error) == nw_error_domain_posix && nw_error_get_error_code(error) == ECANCELED))
-                session->streamReceiveBytes(protectedThis->m_identifier, { }, false, WebCore::Exception(WebCore::ExceptionCode::NetworkError));
+            if (!(nw_error_get_error_domain(error) == nw_error_domain_posix && nw_error_get_error_code(error) == ECANCELED)) {
```
I think using De Morgan's law to remove the parentheses would make this read a little better.

Commit:
WebTransport should be able to cancel reads or writes for a stream
https://bugs.webkit.org/show_bug.cgi?id=303152
rdar://160150325

Reviewed by Alex Christensen.

This PR plumbs sending and receiving RESET_STREAM_AT and STOP_SENDING signals on the streams.

The sending flow does not take the stream error code into account right now.
This is tracked by existing FIXMEs in `WebTransportReceiveStreamSource::doCancel` and `WebTransportSendStreamSink::error`.

It also fixes the cleanup procedure for WebTransport, using the correct exception to terminate the readable and writable streams.

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/WebTransport.mm

* Source/WebCore/Modules/streams/ReadableStreamSource.cpp:
(WebCore::ReadableStreamSource::error):
* Source/WebCore/Modules/streams/ReadableStreamSource.h:
* Source/WebCore/Modules/streams/WritableStream.cpp:
(WebCore::WritableStream::errorIfPossible):
* Source/WebCore/Modules/streams/WritableStream.h:
* Source/WebCore/Modules/webtransport/DatagramByteSource.cpp:
(WebCore::DatagramByteSource::error):
* Source/WebCore/Modules/webtransport/DatagramByteSource.h:
* Source/WebCore/Modules/webtransport/DatagramSource.cpp:
(WebCore::DatagramDefaultSource::error):
* Source/WebCore/Modules/webtransport/DatagramSource.h:
* Source/WebCore/Modules/webtransport/WebTransport.cpp:
(WebCore::WebTransport::receiveBidirectionalStream):
(WebCore::WebTransport::streamReceiveError):
(WebCore::WebTransport::streamSendError):
(WebCore::WebTransport::cleanup):
(WebCore::WebTransport::createBidirectionalStream):
(WebCore::WebTransport::createUnidirectionalStream):
(WebCore::WebTransport::didFail):
* Source/WebCore/Modules/webtransport/WebTransport.h:
* Source/WebCore/Modules/webtransport/WebTransportReceiveStreamSource.cpp:
(WebCore::WebTransportReceiveStreamSource::receiveError):
* Source/WebCore/Modules/webtransport/WebTransportReceiveStreamSource.h:
* Source/WebCore/Modules/webtransport/WebTransportSessionClient.h:
* Source/WebCore/Modules/webtransport/WorkerWebTransportSession.cpp:
(WebCore::WorkerWebTransportSession::didFail):
(WebCore::WorkerWebTransportSession::streamReceiveError):
(WebCore::WorkerWebTransportSession::streamSendError):
* Source/WebCore/Modules/webtransport/WorkerWebTransportSession.h:
* Source/WebCore/PAL/pal/spi/cocoa/NetworkSPI.h:
* Source/WebCore/bindings/js/InternalWritableStream.cpp:
(WebCore::InternalWritableStream::errorIfPossible):
* Source/WebCore/bindings/js/InternalWritableStream.h:
* Source/WebCore/bindings/js/ReadableStreamDefaultController.cpp:
(WebCore::ReadableStreamDefaultController::error):
* Source/WebCore/bindings/js/ReadableStreamDefaultController.h:
* Source/WebKit/Configurations/AllowedSPI.toml:
* Source/WebKit/NetworkProcess/cocoa/NetworkSoftLink.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkSoftLink.mm:
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.cpp:
(WebKit::NetworkTransportSession::streamReceiveError):
(WebKit::NetworkTransportSession::streamSendError):
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.h:
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportStream.h:
* Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
(WebKit::NetworkTransportSession::setupConnectionHandler):
(WebKit::NetworkTransportSession::createStream):
* Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportStreamCocoa.mm:
(WebKit::NetworkTransportStream::sendBytes):
(WebKit::NetworkTransportStream::receiveLoop):
(WebKit::NetworkTransportStream::cancel):
(WebKit::NetworkTransportStream::cancelReceive):
(WebKit::NetworkTransportStream::cancelSend):
(WebKit::NetworkTransportStream::setErrorCodeForStream): Deleted.
* Source/WebKit/WebProcess/Network/WebTransportSession.cpp:
(WebKit::WebTransportSession::streamReceiveError):
(WebKit::WebTransportSession::streamSendError):
(WebKit::WebTransportSession::didFail):
* Source/WebKit/WebProcess/Network/WebTransportSession.h:
* Source/WebKit/WebProcess/Network/WebTransportSession.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebTransport.mm:
(TestWebKitAPI::TEST(WebTransport, NetworkProcessCrash)):

Canonical link: https://commits.webkit.org/303774@main
